File: frontend/src/game/entities/geometry/line-2d.ts
import { Vector2Entity } from './vector2-entity';

let numerator: number;

export class Line2D {
    protected readonly xCoefficient: number;

    protected readonly yCoefficient: number;

    protected readonly constant: number;

    private readonly distanceCoefficient: number;

    constructor(from: Vector2Entity, to: Vector2Entity) {
        // To derive these lines of code, write down the equations "Ax1 + By1 + C = 0" and "Ax2 + By2 + C = 0" and then
        // subtract the first one from the second one, solve for B and then make A equal "y1 - y2" so that it cancels
        // out with the denominator in B's equation.
        this.xCoefficient = from.y - to.y;
        this.yCoefficient = to.x - from.x;
        this.constant = -(this.xCoefficient * from.x + this.yCoefficient * from.y);

        this.distanceCoefficient =
            1.0 / (this.xCoefficient * this.xCoefficient + this.yCoefficient * this.yCoefficient);
    }

    distanceToPointSquared(x: number, y: number): number {
        // Derived with a bunch of trigonometry. Surprisingly graceful equation.
        numerator = this.xCoefficient * x + this.yCoefficient * y + this.constant;
        return numerator * numerator * this.distanceCoefficient;
    }

    solveForPoint(x: number, y: number): number {
        return this.xCoefficient * x + this.yCoefficient * y + this.constant;
    }
}

File: core/src/main/java/moe/mewore/rabbit/geometry/Line2D.java
package moe.mewore.rabbit.geometry;

public class Line2D {

    protected final double xCoefficient;

    protected final double yCoefficient;

    protected final double constant;

    private final double distanceCoefficient;

    Line2D(final Vector2 from, final Vector2 to) {
        // To derive these lines of code, write down the equations "Ax1 + By1 + C = 0" and "Ax2 + By2 + C = 0" and then
        // subtract the first one from the second one, solve for B and then make A equal "y1 - y2" so that it cancels
        // out with the denominator in B's equation.
        xCoefficient = from.getY() - to.getY();
        yCoefficient = to.getX() - from.getX();
        constant = -(xCoefficient * from.getX() + yCoefficient * from.getY());

        distanceCoefficient = 1.0 / (xCoefficient * xCoefficient + yCoefficient * yCoefficient);
    }

    public double distanceToPointSquared(final Vector2 point) {
        // Derived with a bunch of trigonometry. Surprisingly graceful equation.
        final double numerator = xCoefficient * point.getX() + yCoefficient * point.getY() + constant;
        return numerator * numerator * distanceCoefficient;
    }

    public double solveForPoint(final Vector2 point) {
        return xCoefficient * point.getX() + yCoefficient * point.getY() + constant;
    }
}
